lib/plug/upload.ex

```diff
@@ -169,6 +233,21 @@ defmodule Plug.Upload do
     {:reply, {:ok, dirs}, dirs}
   end
 
+  def handle_call(:roots, _from, dirs) do
+    {:reply, {:ok, dirs}, dirs}
+  end
+
+  def handle_call({:give_away, pid, tmp, path}, _from, dirs) do
+    # In the case of accepting a file from another process we need to be sure
+    # that the given file is definitely monitored before removing the original
+    # reference in the @path_table
```

Comment:
It is not about being sure it was monitored before transferring - we could enforce by calling monitor before calling ets - but it is rather about making sure the server won't collect the DOWN message before we add the path to the table. In other words, the three operations below need to happen atomically:

```suggestion
    # Since we are writing in behalf of another process, we need to make sure
    # the monitor and writing to the tables happen within the same operation.
```

Reply:
Much better, thanks.